Set module resolution based on baseUrl in jsconfig/tsconfig.json

rovansteen opened the pull request:
~Based on [this thread](https://twitter.com/dan_abramov/status/1080821938224857088) on Twitter here is a PR to kickstart development/discussion on supporting absolute paths in Typescript via NODE_PATH so they work consistently between JS & TS.~

Updated: Based on the discussion here: https://github.com/facebook/create-react-app/issues/5645#issuecomment-451170386 This attempts to unify the configuration by reading `baseUrl` from `jsconfig.json` / `tsconfig.json`.

🚧 **This PR is still in progress**

To do:
- [x] Add tests
- [ ] Make sure paths are valid TypeScript
- [ ] Add deprecation message to `react-scripts start`
- [ ] Add documentation

gaearon commented:
`@` is likely a larger chunk of work so it's better to do separately.

We'll need to test Windows.

---

While we're at it, I also have an alternative proposal: https://github.com/facebook/create-react-app/issues/5645#issuecomment-451170386. What do you think about it?

rovansteen commented:
I'm having some difficulties with the tests both locally and on the CI. Anyone that has some experience with the test suite in CRA that has some time to help me out?

IIFE commented on packages/react-scripts/config/config.js:
```diff
@@ -0,0 +1,81 @@
+    return null;
+  }
+
+  if (baseUrl !== 'src' && baseUrl !== './src') {
```
Should we restrict it to only `src` (or `./src`), or should we only say that `baseUrl` has to **start** with `src`? Maybe there is a need to have the `baseUrl` start from a directory nested inside `src`.

rovansteen replied:
Right now we only want to support `src`, if anyone has a good use case for setting the baseUrl to a nested directory we can always change it later.

IIFE commented on packages/react-scripts/config/config.js:
```diff
@@ -0,0 +1,81 @@
+    return null;
+  }
+
+  return 'src';
```
Returning `src` regardless of user choice. What if user chose `./src` instead?

rovansteen replied:
`src` is the same as `./src` but by always returning `src` we know what we are working with when passing the value to webpack/jest.

IIFE commented on packages/react-scripts/config/config.js:
```diff
@@ -0,0 +1,81 @@
+ *
+ * @param {Object} options
+ */
+function getPaths(options = {}) {
```
Do we need that? I cannot seem to see the `paths` value being used at all. Also you're not returning `paths` if it's defined, the function currently either returns an empty array of undefined.

rovansteen replied:
This is still work in progress to support paths/aliases but I wanted to get the test working first.

IIFE commented:
@rovansteen I've just added few comments to the changes (first time reviewing open source code :) ).

Hope they make sense.

rovansteen commented:
@IIFE thanks for your feedback! :D

mrmckeb commented:
@rovansteen, if we're deprecating `NODE_PATH`, perhaps we should look at adding some warning here - or migration steps?

rovansteen commented:
@mrmckeb fair enough. Got any ideas about warnings or migration steps?

mrmckeb commented:
@rovansteen I'd just check if NODE_PATH was passed in, and if it was, add a warning to the console and mention that it's been deprecated in favour of using either a `jsconfig` or `tsconfig` file to achieve the same thing - and it would be great to link through to any related docs.

The concern here is that some people (including me) already rely on `NODE_PATH`, and might not read the release notes carefully enough when updating to a new `react-scripts` version.

Timer commented on packages/react-scripts/config/config.js:
```diff
@@ -0,0 +1,110 @@
+const chalk = require('chalk');
+const paths = require('./paths');
+
+const appDirectory = fs.realpathSync(process.cwd());
```
Anywhere that uses this should probably use `paths.appPath` instead.

rovansteen replied:
Implemented!

Timer commented on packages/react-scripts/config/config.js:
```diff
@@ -0,0 +1,110 @@
+    return null;
+  }
+
+  if (baseUrl !== 'src' && baseUrl !== './src') {
```
This needs to support `node_modules` as well, which is the "default value".

You cannot set `paths` without `baseUrl`, so users who want to configure only aliases will want to keep the default behavior of `node_modules`.

Timer replied:
https://www.typescriptlang.org/docs/handbook/module-resolution.html#path-mapping

![image](https://user-images.githubusercontent.com/616428/50743949-9f5a6900-11eb-11e9-8a28-dffac65565f8.png)

mrmckeb replied:
For reference, this is how I have this configures on a fork of react-scripts (for a client project) - it works well, and let's me import (as an example) `../../utils/xyz` as `utils/xyz`.
```
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "*": [
        "src/*"
      ]
    }
  },
```

rovansteen replied:
Hmm, the current implementation would not support syntax like `src/*`. Is there any differences between that and just `src`?

mrmckeb replied:
If you remove the `/*` TypeScript will fail to resolve. However, as we're currently supporting `src`-only (as I understand it) the config you've provided is good.

If we supported other directories, we could use an approach like this:
```
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "some-module": ["not-src/some-module"]
      "*": ["src/*"],
      "@*": ["src/*"]
    }
  },
```

rovansteen replied:
Hmm, but we'll want consistent behavior between editor & CRA so we should probably support this `/*`.

Timer commented on packages/react-scripts/config/config.js:
```diff
@@ -0,0 +1,110 @@
+    return null;
+  }
+
+  if (baseUrl !== 'src' && baseUrl !== './src') {
```
Additionally, we probably shouldn't check the value as a string -- let's resolve the path and verify it's either `src` or `node_modules` (that way we support `./src/`, `./src`, `src/`, `.\src`, etc).

rovansteen replied:
Implemented!

Timer commented on packages/react-scripts/config/config.js:
```diff
@@ -0,0 +1,110 @@
+ *
+ * @param {Object} options
+ */
+function getAlias(options = {}) {
```
We probably don't want to be restrictive about what users can define as an alias.

Since this file is a standard, I'd go so far as to say users can configure any alias(es) they desire. Let's open this up?

Timer replied:
Can you please point me to where @gaearon said this?

And I'm only suggesting we open up the alias part, not the directory.

rovansteen replied:
@Timer can't find it anymore. I asked for his opinion on this specific subject. For now, I've updated the implementation to allow every alias but restrict the path to only allow `src`.

Timer commented on packages/react-scripts/config/config.js:
```diff
@@ -0,0 +1,110 @@
+function getConfig() {
+  // Check if TypeScript is setup
+  const useTypeScript = fs.existsSync(paths.appTsConfig);
+  const hasJsConfig = fs.existsSync(paths.appJsConfig);
```
We should probably enforce not allowing `jsconfig.json` and `tsconfig.json` both being present in a preflight check.

rovansteen replied:
Done!

Timer commented on packages/react-scripts/config/env.js:
```diff
@@ -48,21 +47,14 @@ dotenvFiles.forEach(dotenvFile => {
-  .filter(folder => folder && !path.isAbsolute(folder))
-  .map(folder => path.resolve(appDirectory, folder))
-  .join(path.delimiter);
+if (process.env.NODE_PATH) {
```
We should move this check elsewhere imo -- probably in the script bin wrapper.

rovansteen replied:
You mean this file? https://github.com/facebook/create-react-app/blob/master/packages/react-scripts/bin/react-scripts.js

Timer replied:
Yup!

rovansteen replied:
@Timer in the bin wrapper the `.env` files are not parsed yet, so that fails to check the `NODE_PATH` properly. What's your concern with having it in this file?

Timer replied:
Ah, good point. And this just seems like an odd place to have the check (as a side effect of the file being evaluated).

Timer replied:
Maybe we can just add it to `start.js`, omitting it from the `build` and `test` scripts.

iansu commented on packages/react-scripts/config/webpack.config.js:
```diff
@@ -30,6 +30,8 @@ const paths = require('./paths');
 const ModuleNotFoundPlugin = require('react-dev-utils/ModuleNotFoundPlugin');
 const ForkTsCheckerWebpackPlugin = require('fork-ts-checker-webpack-plugin-alt');
+const config = require('./config');
+console.log(JSON.stringify(config));
```
This line should be removed.

ianschmitz commented on packages/react-scripts/config/webpack.config.js:
```diff
@@ -629,6 +629,7 @@ module.exports = function(webpackEnv) {
             isolatedModules: true,
             noEmit: true,
             jsx: 'preserve',
+            baseUrl: config.baseUrl,
```
We probably can leave this until we do the paths PR, but i think we should start using the app's `tsconfig.json` instead of hard-coding the values here. This would support `paths` and `baseUrl` implicitly.

Timer replied:
Why is this required? I'm not sure I understand.

rovansteen replied:
In my first iteration, I was only showing an error when the `baseUrl` didn't match our requirements and "ignore" it for the webpack configuration. Now we throw an actual error and prevent the rest of the build process we no longer have to overwrite it here.

Timer replied:
Ah, OK. We can probably remove this then. 😄

Timer commented on packages/react-scripts/config/modules.js:
```diff
@@ -0,0 +1,145 @@
+ *
+ * @param {Object} options
+ */
+function getAliases(options = {}) {
```
This function still seems a bit restrictive -- can we allow multiple entries?

rovansteen replied:
It does support multiple entries, e.g.
```json
{
  "compilerOptions": {
    "baseUrl": "src",
    "paths": {
      "@": ["src"],
      "~": ["src"]
    }
  }
}
```

Timer replied:
Ah, we should let them map it to more than just `src`.

rovansteen replied:
Ah misunderstood you then. But only `src` or subfolders of `src` then? As in `path.relative(paths.appSrc, aliasPath) === '' || path.relative(paths.appSrc, aliasPath).startsWith('./')`

Timer replied:
Yeah, subfolders of `src` or `src` itself.

rovansteen commented:
So as @mrmckeb correctly pointed out the current implementation in this PR is not actually valid TypeScript and fails if you run `tsc`. A path (alias) should actually look like this: `"@*": ["src/*"]` rather than this: `"@": ["src"]`.

I'm working on an implementation that supports this but @Timer is there any way we can run `tsc` on the `test/fixtures/typescript` folder, to make sure that even though it builds properly it's also completely valid TypeScript? The whole test setup seems a bit daunting to me so not sure where I should put that.

mrmckeb commented on packages/react-scripts/fixtures/kitchensink/jsconfig.json:
```diff
@@ -0,0 +1,8 @@
+  "compilerOptions": {
+    "baseUrl": "src",
+    "paths": {
+      "@": ["src"]
```
Perhaps add a test with `*` along with `@*` for compatibility with the previous `NODE_PATH` approach.

rovansteen replied:
For people that had their `NODE_PATH` set to `.` and were importing like this: `src/components/button`?

mrmckeb replied:
So, I think it should be the below as a default, as that matches the `NODE_PATH` behaviour. Assuming the user set `NODE_PATH` to `src` would make this a good default:
```
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "*": ["src/*"],
    }
  },
```
Then, a user could do the following, if they chose.
```
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "@*": ["src/*"],
    }
  },
```

rovansteen replied:
Hmm, I'm not sure if we should provide a default as the `NODE_PATH` is not set to `.` by default which would potentially confuse users thinking that's default node behavior.

mrmckeb replied:
I understand, by default there's no NODE_PATH. However, I think most people that use it would use it with `src`.

As per most implementations, you can leave `"baseUrl"` as `"."`, and the aliasing itself can happen in paths.

mrmckeb commented:
Hi @rovansteen, I think @Timer would be better placed to suggest this - as he's been with the project a lot longer than I. 

However, I think you would probably want to add a fixture here:
https://github.com/facebook/create-react-app/tree/master/test/fixtures

Essentially, you'd be testing that the app picks up the configs as expected.

rovansteen commented:
@mrmckeb I have added that, however, there are so many possible configurations we should support it's not feasible to run an integration test for every possible configuration. I'd rather have an integration test to check if the config is picked up as expected (I already added these) and then a bunch of unit tests that just test if the values of the configuration are parsed correctly.

eps1lon commented:
As far as I can tell this would change the behavior of `paths` explained in Microsoft/TypeScript#10866. In depth comment: https://github.com/facebook/create-react-app/issues/5645#issuecomment-460035552

rovansteen commented:
@eps1lon good point, thanks for bringing that up.

@gaearon you've been an advocate for adding some sort of alias  for absolute imports. Do you still want to add that to CRA even if TypeScript's support to this is limited to just compilation?

rovansteen commented:
Since this PR became a bit too big and ambitious by trying to support both baseUrl and aliases I've created a new PR just for supporting baseUrl (#6656). Giving us more room to figure out how to add support for aliases in a stable way.
